keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -0,0 +1,305 @@
+	// for a given mediaType in apiserver.
+	encoder runtime.Encoder
+	// selfLink of the serialized object identifying endpoint
+	// (e.g. group, version, subresource)
+	selfLink string
+}
+
+// TODO: Better name is welcome :)
+type serializationResult struct {
+	once sync.Once
+
+	// raw is serialized object.
+	raw []byte
+	// err is error from serialization.
+	err error
+}
+```
+
+In the initial attempt, watchCache when receiving an event via watch from
+etcd will be opaquing it into `CachingObject` and operating on object of
+that type later.
+
+That means that we won't have gains from avoid serialization for any GET/LIST
+requests server from cache as well as for `init event` that we process when
+initializing a new watch, but that seems good enough for the initial attempt.
+The obvious gain from it is that the memory used for caching is used only
+for a very short period of time (when delivering this watch to watchers) and
+quickly released, which means we don't need to be afraid about increased
+memory usage.
+We may want to revisit that decision later if we would need more gains
+from avoiding serialization and deep-copies of objects in watchcache.
+
+Note that based on a [POC][] (slightly different that above design),
+the gains of implementing it include:
+- eliminating kube-apiserver unresponsiveness in case of write of
+a single huge Endpoints object: [#75294#comment-472728088][]
+- ~7% lower cpu-usage
+- XX% less memory allocations
+
+[POC]: https://github.com/kubernetes/kubernetes/pull/60067
+[#75294#comment-472728088]: https://github.com/kubernetes/kubernetes/issues/75294#issuecomment-472728088
+
+### Risks and Mitigations
+
+Given this proposal doesn't introduce any user visible change, the only risk
+are bugs in implementation. Fortunately, the serialization code is widely
+used by all end-to-end tests, so any bugs should be catched by those or
+unit tests of newly added logic.
+
+## Design Details
+
+### Test Plan
+
+* Unit tests covering all corner cases.
+* Regular e2e tests are passing.
+
+### Graduation Criteria
+
+* All existing e2e tests are passing.
+* Scalability tests confirm gains of that change.
+
+Given it is purely implementation detail, we are not going to expose it
+to users via flag or feature-gate (though the implementation will be done
+behind some kind of hard-coded guard).
+
+### Upgrade / Downgrade Strategy
+
+Not applicable
+
+### Version Skew Strategy
+
+Not applicable
+
+## Implementation History
+
+- 2019-03-27: Created
+
+## Alternatives
+
+### Bake-in caching objects into apimachinery
+We considered making objects above part of apimachinery.
+
+Pros:
+- Expose ability to use it for others
+
+Cons:
+- Complicated code hits apimachinery
+
+### LRU cache
+We considered using simple LRU cache to store serialized objects.
+
+Pros:
+- performance gains also for reads served from etcd (though these
+doesn't seem to be huge based on experiments)
+
+Cons:
+- potentially significant point of contention
```

Comment:
Really?  Isn't there a significant time/space locality to serializations?  Couldn't we be exploiting this affinity between multiple watches and the serialization of an object better?

Reply:
I did a quick and super dirty poc for that an year ago, and the contention was quite huge actually...